torch/csrc/distributed/rpc/utils.cpp

```diff
@@ -713,6 +713,85 @@ std::vector<at::IValue> readWrappedPayload(
   payload.resize(payload.size() - additionalPayloadSize);
   return tupleElements;
 }
+
+void populateRemoteProfiledEvents(
+    std::vector<torch::autograd::profiler::Event>& profiledEvents,
+    const torch::autograd::profiler::ProfilerConfig& profilingConfig,
+    const std::vector<std::vector<torch::autograd::profiler::Event>>&
+        event_lists) {
+  // Gather all events into a vector
+  for (auto& l : event_lists) {
+    for (auto& e : l) {
+      profiledEvents.push_back(e);
+    }
+  }
+  // find __start_profile event and __cuda_start_event.
+  bool cuda_profiling_enabled =
+      profilingConfig.state == torch::autograd::profiler::ProfilerState::CUDA;
+  bool found_cpu_start = false;
+  const torch::autograd::profiler::Event* profilerStart = nullptr;
+  // Each device has its own cudaProfilerStart, so we must take
+  // care to use the correct one depending on the device the
+  // operation ran on.
+  std::unordered_map<int, const torch::autograd::profiler::Event*>
+      cudaProfilerStarts;
+  for (auto& e : profiledEvents) {
+    if (!found_cpu_start && 0 == strcmp(e.name(), "__start_profile")) {
+      profilerStart = &e;
+      found_cpu_start = true;
+    }
+    if (cuda_profiling_enabled && 0 == strcmp(e.name(), "__cuda_start_event")) {
+      e.setCudaUs(e.cpu_us());
+      auto device = e.device();
+      TORCH_CHECK(
+          device != -1,
+          "CUDA profiling was enabled but could not find CUDA device.");
+      TORCH_CHECK(
+          cudaProfilerStarts.find(device) == cudaProfilerStarts.end(),
+          c10::str("Duplicate __cuda_start_event found for ", device));
+      cudaProfilerStarts[device] = &e;
+    }
+    // TODO: determine no. of CUDA devices and break here if we have
+    // a cudaProfilerStart for all of them, in the case of cuda
+    // profiling.
+    if (found_cpu_start && !cuda_profiling_enabled) {
+      break;
+    }
+  }
+  // We should always find __start_profile.
+  TORCH_CHECK(
+      profilerStart != nullptr, "Expected to find __start_profile event.");
+  // Should have >= 1 CUDA start event.
```

Comment:
is this true even if the user function is CPU only but on a machine with CUDA devices? Or is it true that in this case `cuda_profiling_enabled` will be false?

Reply:
If the user function is CPU-only but on a CUDA machine it should still be true since we push `__cuda_start_event` regardless if running with `profile_cuda` flag. Although this is a good point, let me check what happens when we enable this on a CPU only machine (I think we should crash during profiler creation probably)

Reply:
On the caller if there's no CUDA device, then we fail as expected with a message that cuda is not enabled.

On the callee, we get a stacktrace with the following error message:
```
> W0918 11:52:41.866964 1090162 record_function.cpp:171] Exception in RecordFunction callback: CUDA used in profil
er but not enabled.
```

That can happen if machine A sends an RPC to B and A has CUDA but B does not. I wonder if in this case, instead of crashing if we want to support this, it's worth it to check if B  also has CUDA, and fallback to CPU-only profiler in the case that it does not instead of crashing?

Reply:
https://github.com/pytorch/pytorch/pull/44967

Reply:
> I wonder if in this case, instead of crashing if we want to support this

Yep, I would vote for supporting this. I do saw users on forum who wants to connect CUDA with non-CUDA servers using RPC.